EOS S3 SPI driver

arturkow2 opened the pull request.

arturkow2000 committed:
drivers: spi: eos_s3: Add driver for EOS S3

Add SPI driver and bindings for SPI controller on EOS S3 SOC. Based on
driver code from Antmicro's Zephyr fork.

Signed-off-by: Artur Kowalski <artur.kowalski@3mdeb.com>

kartben commented:
Thanks for the PR! You may want to put it in "draft" mode while it's still wip so that reviewers only look at it when it's ready. Thanks!

arturkow2000 committed:
soc: arm: eos_s3: disable A1 clock divisor

SPI sources its clock from A1 domain, ensure the clock is configured
accordingly after reset.

fkokosinski commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,365 @@
+	uint32_t clock_div = SPI_CLK / config->frequency;
+
+	/* Divisor must be an even number */
+	if (clock_div % 2) {
+		clock_div++;
+	}
+
+	/* This driver doesn't work correctly with clock rate higher than 2MHz.
+	 * 10MHz / 6 ~= 1.6 MHz
+	 */
+	if (clock_div < 6)
+		return -EINVAL;
```
Do we need to always increment `clock_div` when it's not an even divisor? What if `clock_div` was 5, got incremented to 6 instead of decremented to 4, and then rejected because it's not `< 6`?

arturkow2 replied:
> Do we need to always increment clock_div when it's not an even divisor?

This is what the original driver from Antmicro's fork did. I've checked EOS S3 docs, and indeed it is possible to divide the clock only by multiples of 2 (and 2 is the minimum divisor).

> What if clock_div was 5, got incremented to 6 instead of decremented to 4, and then rejected because it's not < 6?

Driver rejects divisors smaller than 6, not bigger.

fkokosinski replied:
Fair enough, thanks for elaborating on it

arturkow2 commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,365 @@
+#include <eoss3_hal_spi.h>
+#include <eoss3_dev.h>
+#include "spi_eos_s3.h"
+#if defined(CONFIG_PINCTRL)
```
done

arturkow2 commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,365 @@
+
+struct spi_eos_s3_config {
+	SPI_TypeDef *regs;
+#if defined(CONFIG_PINCTRL)
```
done

arturkow2 commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,365 @@
+
+static int spi_eos_s3_init(const struct device *dev)
+{
+#if defined(CONFIG_PINCTRL)
```
done

arturkow2 commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,365 @@
+{
+#if defined(CONFIG_PINCTRL)
+	int ret;
+	const struct spi_eos_s3_config *config = dev->config;
```
done

arturkow2 commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,365 @@
+	int tx_frame;
+	int rx_frame;
+
+	/* Do not send data when continous RX mode is selected */
```
done

arturkow2 commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,365 @@
+	SPI_TypeDef *regs = cfg->regs;
+
+	if (data->mode == RX_MODE) {
+		/* Set number of bytes for continous read */
```
done

arturkow2 commented on soc/arm/quicklogic_eos_s3/soc.c:
```diff
@@ -39,6 +39,7 @@ static void eos_s3_cru_init(void)
 
 	/* Turn off divisor for A0 domain */
 	CRU->CLK_CTRL_A_0 = 0;
+	CRU->CLK_CTRL_A_1 = 0;
```
done

arturkow2 commented on include/zephyr/dt-bindings/pinctrl/quicklogic-eos-s3-pinctrl.h:
```diff
@@ -21,5 +21,9 @@
+#define SPI_CLK_PAD34		QUICKLOGIC_EOS_S3_PINMUX(34, 0x00)
+#define SPI_MISO_PAD36		QUICKLOGIC_EOS_S3_PINMUX(36, 0x124 << 20)
+#define SPI_MOSI_PAD38		QUICKLOGIC_EOS_S3_PINMUX(38, 0x00)
+#define SPI_SS1_PAD39		QUICKLOGIC_EOS_S3_PINMUX(39, 0x00)
```
done

fkokosinski commented:
hi @arturkow2, thanks for addressing my comments!

There is one additional thing - compliance checks job failed: https://github.com/zephyrproject-rtos/zephyr/actions/runs/5928643323/job/16095193804?pr=60841. Could you take a look at it?

tbursztyka commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,353 @@
+	ret = spi_eos_s3_configure(regs, config, data->mode);
+
+	if (ret) {
+		spi_context_release(&data->ctx, ret);
```
Btw, you need to call spi_context_lock() at the beginning to this transceive function

tbursztyka commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,353 @@
+static int spi_eos_s3_release(const struct device *dev,
+			     const struct spi_config *config)
+{
+	const struct spi_eos_s3_config *eos_s3_cfg = dev->config;
```
you have to check if given config is the one being in use atm. (spi_context_configured() is your friend here).

and you need to call, at the end of the function spi_context_unlock_unconditionally()

tbursztyka commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,353 @@
+	} while (!ret && spi_eos_s3_transfer_ongoing(data));
+
+	spi_eos_s3_finish(regs);
+	spi_context_complete(ctx, dev, 0);
```
So it's a polling transfer function. 
The controller does not support interruption?

tbursztyka commented on drivers/spi/spi_eos_s3.c:
```diff
@@ -0,0 +1,353 @@
+
+	rx_frame = spi_eos_s3_rx(regs);
+
+	if (rx_frame == NO_RX_DATA)
```
even one liner if statements need {} per-zephyr code style (apply that where relevant)
